rna/unafold/doc/hybrid-select.pl.1.in
.TH HYBRID-SELECT.PL 1 "@DATE@" "@PACKAGE_NAME@ @PACKAGE_VERSION@" "User Commands"
.SH NAME
hybrid-select.pl \- fold or hybridize RNA or DNA sequences with selected base pairs
.SH SYNOPSIS
.B hybrid-select.pl
.RI [ OPTION ]...
.I FILE1
.RI [ FILE2 ]
.SH DESCRIPTION
\fBhybrid-select.pl\fR performs two folding or hybridization calculations in sequence, with the output of the first calculation used to create a list of valid abse pairs for the second calculation.
.P
Both the initial and final calculations may be either energy minimization or partition function; the default is an initial partition function calculation followed by an energy minimization.
.P
When the initial calculation is a partition function, only base pairs having a specified probability of forming are retained for the second run.  If the filtering stage is an energy minimization, only base pairs appearing in structures with energies within a specified percentage of the minimum energy are retained.
.P
If the final calculation is an energy minimization, the \fB--mfold\fR option may be used; likewise, \fB--tracebacks\fR is acceptable if the final calculation is a partition function.
.SH OPTIONS
Most of the options below are passed to the subprograms to which they apply, but \fB--cutoff\fR, \fB--initial\fR and \fB--final\fR affect the behavior of \fBhybrid-select.pl\fR directly.
.TP
.BR -n ", " --NA =RNA|DNA
sets nucleic acid type to RNA or DNA.  Default is RNA.
.TP
.BR -t ", " --t =\fItemp
sets temperature to \fItemp\fR.  Default is 37.
.TP
.BR -N ", " --sodium =\fIvalue
sets Sodium ion concentration to \fIvalue\fR molar.  Default is 1.
.TP
.BR -M ", " --magnesium =\fIvalue
sets Magnesium ion concentration to \fIvalue\fR molar.  Default is 0.
.TP
.BR -p ", " --polymer
use salt corrections for polymers instead of oligomers (the default).
.TP
.BR -s ", " --suffix =\fIstring
use energy rules with the suffix \fIstring\fR.  Overrides \fB--tmin\fR, \fB--tinc\fR, \fB--tmax\fR, \fB--sodium\fR, \fB--magnesium\fR and \fB--polymer\fR.
.TP
.BR -r ", " --prohibit =\fIi\fR,\fIj\fR[,\fIk\fR]
prohibit all basepairs in the helix from \fIi\fR,\fIj\fR to \fIi+k-1\fR,\fIj-k+1\fR.  If \fIj\fR is 0, prohibit bases \fIi\fR to \fIi+k-1\fR from pairing at all; if \fIi\fR is 0, prohibit bases \fIj\fR to \fIj-k+1\fR from pairing at all.  \fIk\fR defaults to 1 if not specified.
.TP
.BR -f ", " --force =\fIi\fR,\fIj\fR[,\fIk\fR]
force all basepairs in the helix from \fIi\fR,\fIj\fR to \fIi+k-1\fR,\fIj-k+1\fR.  If \fIj\fR is 0, forces bases \fIi\fR to \fIi+k-1\fR to be double-stranded; if \fIi\fR is 0, forces bases \fIj\fR to \fIj-k+1\fR to be double-stranded.  \fIk\fR defaults to 1 if not specified.
.TP
.BR -I ", " --noisolate
prohibits all isolated basepairs.  Isolated basepairs are helices of length 1; that is, they do not stack on another basepair on either side.  (See also the \fB--prefilter\fR and \fB--nopostfilter\fR options below.)
.TP
.BR -k ", " --tracebacks =\fInumber
computes \fInumber\fR stochastic tracebacks and writes them to \fIprefix.temperature.ct\fR.  Stochastic tracebacks are computed according to the Boltzmann probability distribution so that the probability of a structure is its Boltzmann factor divided by the partition function.
.TP
.BR -m ", " --maxbp =\fIvalue
Bases farther apart than \fIvalue\fR cannot form.  Default is no limit.
.TP
.BR -c ", " --constraints =\fIfile
reads a list of constraints from \fIfile\fR.  Constraints must be in the form "P \fIi\fR \fIj\fR \fIk\fR" or "F \fIi\fR \fIj\fR \fIk\fR".  These are equivalent to specifying "--prohibit=\fIi\fR,\fIj\fR,\fIk\fR" and "--force=\fIi\fR,\fIj\fR,\fIk\fR", respectively.  If \fIfile\fR is not specified, it defaults to \fIprefix.aux\fR.
.TP
.BR -C ", " --cutoff =\fIvalue
Sets the cutoff for basepairs to be included in the second run, as a probability or a percent of the minimum energy.  Default is 0.001 (partition function) or 5% (energy minimization).
.TP
.BR --initial =\fIprogram
Specifies the program to run in the first phase.  May be the name of a program like \fBhybrid-ss\fR or one of the codes "PF" and "EM".  PF chooses \fBhybrid-ss\fR or \fBhybrid\fR, while EM chooses \fBhybrid-ss-min\fR or \fBhybrid-min\fR.  Default is PF.
.TP
.BR --final =\fIprogram
Specifies the program to run in the second phase, using the same choices as \fB--initial\fR.  Default is EM.
.SH OBSCURE OPTIONS
.TP
.B --allpairs
allows basepairs to form between any two nucleotides.  When \fB--allpairs\fR is not specified, only Watson-Crick and wobble basepairs are allowed.
.TP
.BR --maxloop =\fIsize
sets the maximum size of bulge/interior loops to \fIsize\fR.  Default is 30.
.TP
.B --nodangle
removes single-base stacking from consideration.
.TP
.B --simple
makes the penalty for multibranch loops constant rather than affine.
.TP
.BR --prefilter =\fIvalue1\fR[,\fIvalue2\fR]
Sets the prefilter to filter out all basepairs except those in groups of \fIvalue2\fR adjacent basepairs of which \fIvalue1\fR can form.  \fIvalue2\fR is the same as \fIvalue1\fR if unspecified.  Default is 2 of 2.  (See also the \fB--noisolate\fR option above.)
.TP
.B --nopostfilter
Disables the postfilter.  The postfilter, which is enabled by default, removes from consideration all structures that consist of only one basepair.
.SH ENVIRONMENT
.IP UNAFOLDDAT
an alternate location from which to read the energy rules.  The default energy rules can be overridden with files in the current directory or in the directory pointed to by UNAFOLDDAT.  \fBhybrid-select.pl\fR looks for each file first in the current directory, then in the directory specified by UNAFOLDDAT and last in \fI/usr/local/share/unafold\fR (or wherever the energy rules were installed).
.SH REFERENCES
@REFERENCES@
.SH "SEE ALSO"
.BR hybrid-ss (1),
.BR hybrid-ss-min (1),
.BR hybrid (1),
.BR hybrid-min (1)
.SH AUTHORS
Nick Markham <markhn@rpi.edu> and Michael Zuker <zukerm@rpi.edu>
.SH COPYRIGHT
Copyright (c) 2006-2009, Rensselaer Polytechnic Institute.
.SH AVAILABILITY
UNAFold is available from http://dinamelt.bioinfo.rpi.edu/unafold/.  Commercial use requires a license; see http://dinamelt.bioinfo.rpi.edu/commercial.php.


rna/unafold/scripts/hybrid-select.pl
#! /opt/local/bin/perl -w

use strict;
use warnings;
use Getopt::Long;
Getopt::Long::Configure 'gnu_getopt', 'no_auto_abbrev', 'no_ignore_case';

sub version ($) {
    print "$_[0] (UNAFold) 3.8\n";
    print "By Nicholas R. Markham and Michael Zuker\n";
    print "Copyright (C) 2006\n";
    print "Rensselaer Polytechnic Institute\n";
    print "Troy, NY 12810-3590 USA\n";
    exit;
}

sub usage () {
    print <<EOF;
Usage: hybrid-select.pl [options] file1 [file2]

Options:
-n, --NA=(RNA | DNA) (defaults to RNA)
-t, --t=<temperature> (defaults to 37)
-s, --suffix=<free energy suffix>
-N, --sodium=<[Na+] in M> (defaults to 1)
-M, --magnesium=<[Mg++] in M> (defaults to 0)
-p, --polymer
-h, --prohibit=<i,j,k>
-f, --force=<i,j,k>
-I, --noisolate
-F, --mfold=<P,W,MAX> (defaults to 5,3,100)
-k, --tracebacks=<number of tracebacks>
-m, --maxbp=<maximum basepair distance>
-c, --constraints=<name of constraint file> (defaults to prefix.aux)
-C, --cutoff=<energy/probability cutoff> (defaults to 5%/0.001)
    --initial=<first program> (defaults to PF)
    --final=<second program> (defaults to EM)

Obscure options:
    --allpairs
    --maxloop=<maximum bulge/interior loop size> (defaults to 30)
    --nodangle
    --simple
    --prefilter=<filter value>
    --nopostfilter
EOF
    print 'Report bugs to markhn@rpi.edu', "\n";
    exit;
}

my ($temp, $cutoff, $prog1, $prog2) = (37, undef, 'PF', 'EM');
my ($mfold, $tracebacks, @prohibit, @force);
my %options;

GetOptions \%options, 'version|V' => sub { version('hybrid-select.pl') }, 'help|h' => sub { usage() }, 'NA|n=s', 't=f' => \$temp, 'suffix|s=s', 'allpairs', 'maxloop=i', 'sodium|N=f', 'magnesium|M=f', 'polymer|p', 'prohibit|r=s' => \@prohibit, 'force|f=s' => \@force, 'nodangle', 'simple', 'prefilter=s', 'nopostfilter', 'energyOnly|E', 'noisolate|I', 'mfold|F:s' => \$mfold, 'tracebacks|k=i' => \$tracebacks, 'maxbp|m=i', 'constraints|c:s', 'cutoff|C=f' => \$cutoff, 'initial=s' => \$prog1, 'final=s' => \$prog2 or die $!;

my @args = ('--tmin' => $temp, '--tmax' => $temp);

foreach (keys %options) {
    if ($_ eq 'allpairs' or $_ eq 'nodangle' or $_ eq 'polymer' or $_ eq 'noisolate'or $_ eq 'simple' or $_ eq 'nopostfilter') {
	push @args, "--$_";
    } else {
	push @args, "--$_";
	push @args, $options{$_} if $options{$_};
    }
}

push @args, '--prohibit' => $_ foreach @prohibit;
push @args, '--force' => $_ foreach @force;

unless ($ARGV[0]) {
    print STDERR "Error: file not specified\nRun 'hybrid-select.pl -h' for help\n";
    exit 1;
}
my ($file1, $file2) = @ARGV;
defined $file2 or $file2 = '';
my ($prefix1, $prefix2);
($prefix1 = $file1) =~ s/\.seq$//;
($prefix2 = $file2) =~ s/\.seq$// if $file2;

my $prefix = $file2 ? "$prefix1-$prefix2" : $prefix1;

my ($seq1, $seq2);
open IN, '<', $file1 or open IN, '<', "$file1.seq" or die $!;
$seq1 = join '', <IN>;
close IN or die $!;
$seq1 =~ tr/A-Za-z0-9//cd;
if ($file2) {
    open IN, '<', $file2 or open IN, '<', "$file2.seq" or die $!;
    $seq2 = join '', <IN>;
    close IN or die $!;
    $seq2 =~ tr/A-Za-z0-9//cd;
}

for ($prog1, $prog2) {
    if ($_ eq 'PF') {
	$_ = $file2 ? 'hybrid' : 'hybrid-ss';
    } elsif ($_ eq 'EM') {
	$_ = $file2 ? 'hybrid-min' : 'hybrid-ss-min';
    }
    if ($_ eq 'hybrid-ss' and $options{nodangle} and $options{simple}) {
	print "Replacing hybrid-ss --nodangle --simple with hybrid-ss-simple\n";
	$_ = 'hybrid-ss-simple';
    } elsif ($_ eq 'hybrid-ss' and length $seq1 <= 11) {
	print "Replacing hybrid-ss with hybrid-ss-noml\n";
	$_ = 'hybrid-ss-noml';
    }
}

print "Warning: ignoring --mfold option\n" if defined $mfold and $prog2 !~ /min/;
print "Warning: ignoring --tracebacks olption\n" if defined $tracebacks and $prog2 =~ /min/;

unless (defined $cutoff) {
    if ($prog1 =~ /min/) {
	$cutoff = 5;
    } else {
	$cutoff = 0.001;
    }
}

system($prog1, ($prog1 =~ /min/ ? '--mfold' : ()), @args, $file1, $file2) == 0 or die $!;

if ($prog1 =~ /min/) {
    open IN, '<', "$prefix.dG" or die $!;
    scalar <IN>;
    $_ = scalar <IN>;
    my (undef, $dG, undef) = split;
    $cutoff = (100 - $cutoff) * $dG / 100;
    close IN or die $!;

    open IN, '<', "$prefix.plot" or die $!;
    open OUT, '>', "$prefix.bp" or die $!;
    scalar <IN>;
    while (<IN>) {
	chomp;
	my (undef, $k, $i, $j, $e) = split or next;
	$i > 0 and $j > 0 and $k > 0 and $e / 10.0 < $cutoff or next;
	$j -= length($seq1) if $file2;
	print OUT "$i\t$j\t$k\n";
    }
    close OUT or die $!;
    close IN or die $!;
} else {
    open IN, '<', "$prefix.$temp.plot" or die $!;
    open OUT, '>', "$prefix.bp" or die $!;
    scalar <IN>;
    while (<IN>) {
	chomp;
	my ($i, $j, $p) = split or next;
	$i > 0 and $j > 0 and $p > $cutoff or next;
	print OUT "$i\t$j\t1\n";
    }
    close OUT or die $!;
    close IN or die $!;
}

if (defined $mfold) {
    push @args, '--mfold';
    $args[$#args] .= '=' . $mfold if $mfold;
}
push @args, '--tracebacks' => $tracebacks if $tracebacks;

system($prog2, @args, '--basepairs' => "$prefix.bp", $file1, $file2) == 0 or die $!;
